packages/scrapers-v2/src/runtime/logger.ts

```diff
@@ -0,0 +1,175 @@
+
+    for (const err of result.err) {
+      if (err instanceof Error) {
+        stats.recordError(err, id);
+      } else {
+        stats.recordError(new Error(`non-error value: ${err}`), id);
+      }
+    }
+  }
+
+  stats.print();
+};
+
+/**
+ * Allows for "recording" fields (and errors) to print a breakdown of the
+ * different values at the end.
+ *
+ * Each field value will be added to a tally. The # of occurrences of each value
+ * for a field are then displayed when `print()` is called.
+ *
+ * Errors are also tallied. Error grouping is done by comparing stacktrace. This
+ * doesn't quite work for async stacktraces, so sometimes two of the same error
+ * are displayed separately.
+ */
+class StatsLogger {
+  // field -> value -> count
+  private fields: Record<string, Map<any, number>> = {};
+  // message -> list -> stacktrace
+  private errors: Map<
+    string,
+    Array<{ err: Error; count: number; annot: string; entryIds: URL[] }>
+  > = new Map();
+
+  /**
+   * Records a field and its value, with the goal of printing the counts for
+   * each different value the field has at the end.
+   *
+   * @param field The name of the field
+   * @param value The value of the field
+   */
+  recordField(field: string, value: any) {
+    if (field === "errors") {
+      throw new Error(
+        "cannot use 'errors' as a field key, use a different name"
+      );
+    }
+    this.record(field, value);
+  }
+
+  /**
+   * Records an error for a specific entry. Error uniqueness is determined by
+   * stack trace. This doesn't quite work for async stacktraces, so sometimes
+   * two of the same error are displayed separately.
+   *
+   * @param err     The error
+   * @param entryId The entry URL
+   */
+  recordError(err: Error, entryId: URL) {
+    const key = err.message ?? "had no stacktrace";
+    const storedErrors = this.errors.get(key) ?? [];
+    for (const stored of storedErrors) {
+      // if the stacktrace matches, increment the count
+      if (err.stack === stored.err.stack) {
+        stored.count += 1;
+        stored.entryIds.push(entryId);
+        this.record("errors", stored.annot);
+        return;
+      }
+    }
+    const id = storedErrors.length === 0 ? "" : ` #${storedErrors.length}`;
+    const annot = `${err.message}${id}`;
+    // stacktrace didn't match, so add a new stacktrace entry for this error message
+    storedErrors.push({ err, count: 1, annot, entryIds: [entryId] });
+    this.errors.set(key, storedErrors);
+    this.record("errors", annot);
+  }
+
+  private record(field: string, value: any) {
+    if (!(field in this.fields)) {
+      this.fields[field] = new Map();
+    }
+    const map = this.fields[field];
+    map.set(value, (map.get(value) ?? 0) + 1);
+  }
+
+  /**
+   * Prints field and error information.
+   *
+   * Prints stacktraces (with URLs) first, then aggregate information. Also
+   * prints in order of # of occurrences.
+   */
+  print() {
```

Comment:
This error logging is sick. I think the general formatting could use a little help since if there are a lot of errors coming out, it can be quite difficult to associate the error msg, # of links that contain as well as the urls. Furthermore, I think it could be really beneficial to abstract where we log them, and not just limit it to the console. Something along the lines of taking in a `StringBuilder` of sorts, and logging all the errors there. Then we can choose whether we want to display this in the CLI or in a text file, or maybe even in an excel sheet for better readability when looking for errors.

Reply:
i agree, it's hard to read with lots of errors. i've listed some additional features that would be nice to have in the [brainstorming doc here](https://www.notion.so/sandboxnu/Scraper-Brainstorming-61261181144c4dac82488255cfa34744#6569d4bd9b9549999b66202686ade507). however, i think these are features for implementing later on, but i totally agree.

in my ideal world, running the `scrape` command opens up a browser tab that has expandable dropdowns, sortable/queryable errors, a database, etc.

Reply:
Damn having another web app for logging of scrapers would be so cool. Hmmmm 🤔 👀